### ISSUE TITLE
`_foreach_add` segfaults when passed tensors with different lengths

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#112305]
### 🐛 Describe the bug

```
torch.ops.aten._foreach_add([torch.randn(5), torch.randn(4)], [torch.randn(3)])
```
=>
segfault

### Versions

N/A

cc @ezyang @gchanan @zou3519 @kadeng

**Comment by Cali0707** [comment]
This only happens when the first list is longer than the second:
```python
import torch
torch._foreach_add([torch.randn(5)], [torch.randn(4), torch.randn(3)])
```
won't segfault

**Comment by malfet** [comment]
It crashes in dispatcher code, where I guess one has to check that lists are of equal size before doing anything:
```
frame #9: 0x00007fffe329833d libtorch_cpu.so`torch::autograd::VariableType::(anonymous namespace)::_foreach_add_List(ks=(repr_ = 274877939713), self=at::TensorList @ 0x00007fffffffc6e0, other=at::TensorList @ 0x00007fffffffc6d0, alpha=0x00007fffffffcb30) at VariableType_0.cpp:1571:82
   1568	  
   1569	  std::vector<bool> _any_has_forward_grad_result(self.size());
   1570	  for (const auto& i : c10::irange(self.size())) {
-> 1571	    _any_has_forward_grad_result[i] = isFwGradDefined(self[i]) || isFwGradDefined(other[i]);
   1572	  }
   1573	  std::shared_ptr<ForeachAddBackward0List> grad_fn;
   1574	  if (_any_requires_grad) {
```

**Comment by malfet** [comment]
Tentatively grabbing for myself, here is the fix, but need to add testing (and perhaps change it to equal):
```python
diff --git a/tools/autograd/gen_variable_type.py b/tools/autograd/gen_variable_type.py
index ed2ca767077..15e9e02a1a3 100644
--- a/tools/autograd/gen_variable_type.py
+++ b/tools/autograd/gen_variable_type.py
@@ -1785,27 +1785,21 @@ def emit_body(
         else:
             for derivative in fw_derivatives:
                 bool_vector_name = get_any_has_forward_grad_name(derivative.var_names)
-                cur_derivative_conditions = [
-                    FW_DERIVATIVE_CHECK_TEMPLATE.substitute(
-                        req_inp=(
-                            inp.name
-                            if not inplace
-                            else refargname2inplace_foreacharg[inp.name].name
-                        )
-                        + (
-                            "[i]"
-                            if is_tensor_list_type(
-                                inp.type
-                                if not inplace
-                                else refargname2inplace_foreacharg[inp.name].type
-                            )
-                            else ""
-                        ),
-                    )
-                    for inp in differentiable_inputs
-                    if derivative.required_inputs_fw_grad is not None
-                    and inp.name in derivative.required_inputs_fw_grad
-                ]
+                cur_derivative_conditions = []
+                for inp in differentiable_inputs:
+                    if derivative.required_inputs_fw_grad is None:
+                        continue
+                    if inp.name not in derivative.required_inputs_fw_grad:
+                        continue
+                    inp_name = inp.name if not inplace else refargname2inplace_foreacharg[inp.name].name
+                    inp_type = inp.type if not inplace else refargname2inplace_foreacharg[inp.name].type
+                    is_list_type = is_tensor_list_type(inp_type)
+                    if is_list_type:
+                        content.append(f"TORCH_CHECK(self.size() <= {inp_name}.size())");
+                        cur_derivative_conditions.append(FW_DERIVATIVE_CHECK_TEMPLATE.substitute(req_inp=inp_name + "[i]"))
+                    else:
+                        cur_derivative_conditions.append(FW_DERIVATIVE_CHECK_TEMPLATE.substitute(req_inp=inp_name))
+
                 content.append(f"std::vector<bool> {bool_vector_name}(self.size());")
                 content.append("for (const auto& i : c10::irange(self.size())) {")
                 content.append(
```

**Comment by zou3519** [comment]
We should also check if the segfault is a regression (i.e. if we didn't crash on this on a previous version of PyTorch)

**Comment by malfet** [comment]
It is a regression, i.e. in 2.0.1 it raised a structured exception:
```
% python3 -c "import torch;print(torch.__version__);print(torch.ops.aten._foreach_add([torch.randn(5), torch.randn(4)], [torch.randn(3)]))"
2.0.1
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/nshulga/miniforge3/envs/py311-torch201/lib/python3.11/site-packages/torch/_ops.py", line 502, in __call__
    return self._op(*args, **kwargs or {})
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Tensor lists must have the same number of tensors, got 2 and 1
```

**Comment by Chillee** [comment]
I wonder what the regression could have been?

**Comment by huydhn** [comment]
The fix is ok for cherry picking 2.1.1, but need to follow up
